Compilation error with anonymous functions inside generic function

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 21.10 (VM)
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-4720HQ CPU @ 2.60GHz
CC version: cc (Ubuntu 11.2.0-7ubuntu2) 11.2.0

getwd: /vagrant
vmodules: /home/vagrant/.vmodules
vroot: /home/vagrant/v
vexe: /home/vagrant/v/v
vexe mtime: 2021-12-01 10:57:12
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 1b691e7.47aa2b1

Git version: git version 2.32.0
Git vroot status: weekly.2021.47-71-g47aa2b1f
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de-dirty

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg repro.v`
```v

fn foo<T>() {
    x := fn() { }
}

fn main() {
    foo<int>()
}

```

**What did you expect to see?**

the compilation succeed


**What did you see instead?**
```
repro.v:3:5: warning: unused variable: `x`
    1 | 
    2 | fn foo<T>() {
    3 |     x := fn() { }
      |     ^
    4 | }
    5 |
==================
/tmp/v_1000/repro.4036707335627059811.tmp.c:10676: error: 'anon_fn_5aacada3bf4819db__29' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error:

doesn't change the result,  if the anon func is defined inside a generic function, invalid C is produced.
if i am not missing something or is a known issue

the following code makes no sense but should show my point:
```v
fn generic_func<T>() fn() {fn generic_func<T>() fn() {
    x := fn() { println('${@FN}') }
    return x
}

fn non_generic_func() fn() {
    x := fn() { println('${@FN}') }
    return x
}

fn main() {
    /* this compiles
    b := non_generic_func()
    b()
    */
    
    f := generic_func<int>()
    f()
}
```

```
/tmp/v_1000/rep.7387829689201297376.tmp.c:10691: error: `anon_fn_f46306f2b135f474__42` undeclared
builder error:
```

in the produced C code we can se that the anon function inside the generic function `foo<T>`  is generated as `anon_fn_f46306f2b135f474__42_T_int` but then referenced in the body of `foo` as `anon_fn_f46306f2b135f474__42`, witout the `T_int` suffix:
```c
[...]

void anon_fn_f46306f2b135f474__42_T_int(void) {
        println( str_intp(2, _MOV((StrIntpData[]){{_SLIT0, 0xfe10, {.d_s = _SLIT("anon_fn_f46306f2b135f474__42")}}, {_SLIT0, 0, { .d_c = 0 }}})));
}

[...]

anon_fn_ main__generic_func_T_int(void) {
        void (*x) (void) =      anon_fn_f46306f2b135f474__42;
        anon_fn_ _t1 = (voidptr)x;
        return _t1;
}

void main__main(void) {
        void (*f) (void) = main__generic_func_T_int();
        f();
}
```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -217,7 +217,9 @@ fn (mut g Gen) gen_fn_decl(node &ast.FnDecl, skip bool) {
 	}
 	mut type_name := g.typ(node.return_type)
 
-	name = g.generic_fn_name(g.cur_concrete_types, name, true)
+	if node.generic_names.len > 0 {
+		name = g.generic_fn_name(g.cur_concrete_types, name, true)
+	}
 
 	if g.pref.translated && node.attrs.contains('c') {
 		// This fixes unknown symbols errors when building separate .c => .v files

--- a/vlib/v/tests/generic_fn_with_anon_fn_test.v
+++ b/vlib/v/tests/generic_fn_with_anon_fn_test.v
@@ -1,0 +1,11 @@
+fn foo<T>() string {
+	x := fn () string {
+		return 'ok'
+	}
+	return x()
+}
+
+fn test_generic_fn_with_anon_fn() {
+	ret := foo<int>()
+	assert ret == 'ok'
+}

